lib/protocol/protocol.go

```diff
@@ -850,17 +862,20 @@ func (c *rawConnection) shouldCompressMessage(msg message) bool {
 func (c *rawConnection) Close(err error) {
 	c.sendCloseOnce.Do(func() {
 		done := make(chan struct{})
-		c.send(&Close{err.Error()}, done)
+		timeout := time.NewTimer(CloseTimeout)
 		select {
-		case <-done:
+		case c.closeBox <- asyncMessage{&Close{err.Error()}, done}:
+			select {
+			case <-done:
+			case <-timeout.C:
+			case <-c.closed:
+			}
+		case <-timeout.C:
 		case <-c.closed:
 		}
 	})
 
-	// No more sends are necessary, therefore further steps to close the
-	// connection outside of this package can proceed immediately.
-	// And this prevents a potential deadlock due to calling c.receiver.Closed
-	go c.internalClose(err)
+	c.internalClose(err)
```

Comment:
Now you're scaring me again by removing anti deadlock stuff without comment...

Reply:
Don't be scared, there's tests now :)

But your right, I should have commented on it: The test added in https://github.com/syncthing/syncthing/pull/5707 checks this deadlock scenario and it is obsolete since closing connections doesn't happen under `model.pmut` lock anymore.